---
title: "R Notebook"
output: word_document
---

#Controls Overview
Two types of controls (extraction and sequencing) were used in this experiment, for a total of 10 different controls:
 Extraction Controls (Artificial Colony, Robogut, Zymobiomics Control 1, Zymobiomics Control2, Extraction Blank)
 Sequencing Controls (ATTC 1-4, Zymobiomics Seq Control)
 
#Libraries
```{r}
#Load required libraries
library("biom")
library(data.table)
```

#Control Makeup
Read in each of the controls composition files
```{r}
control_full <- read.table("T:\\DCEG\\CGF\\TechTransfer\\Microbiome\\Extraction\\Optimization\\Fresh Fecal Optimization_2017.08\\Controls\\Compositions\\MasterSpeciesList.txt",sep="\t",header=TRUE)

#Create Control Composition Database, removing rows of species not in samples, followed by samples not of specified control type
##Artifical Colony
art.col <- subset(control_full, !(Art.Col=='A'))
art.col <- art.col[,1:10]

##Two concentrations of zymo extraction controls
zymo.ext200 <- subset(control_full, !(Zymo.Ext=='A'))
zymo.ext200 <- zymo.ext200[,c(1:9,15)]
zymo.ext2000 <- zymo.ext200

#Zymo Sequencing Control
zymo.seq <-subset(control_full, !(Zymo.Seq=='A'))
zymo.seq <- zymo.seq[,c(1:9,16)]

#ATTC Controls
msa1000 <-subset(control_full, !(MSA1000=='A'))
msa1000 <- msa1000[,c(1:9,11)]

msa1001 <- subset(control_full, !(MSA1001=='A'))
msa1001 <- msa1001[,c(1:9,12)]

msa1002 <- subset(control_full, !(MSA1002=='A'))
msa1002 <- msa1002[,c(1:9,13)]

msa1003 <- subset(control_full, !(MSA1003=='A'))
msa1003 <- msa1003[,c(1:9,14)]

#Create list of all the species to review
species_list_controls <- as.character(unique(control_full$Species))
species_list_controls

#Create list of all the genus to review
genus_list_controls <- as.character(unique(control_full$Genus))
genus_list_controls
```

#Sample Data
Read in the entire sample and OTU databses, then merge the databases. Finally filter for controls only
```{r}
#Read in database with variable information and list of samples that did not meet the sequencing threshold
complete_info <- read.table("T:\\DCEG\\CGF\\TechTransfer\\Microbiome\\Extraction\\Optimization\\Fresh Fecal Optimization_2017.08\\Analysis\\Run1&2\\Manifests\\sampledata.txt",sep="\t",header=TRUE)
complete_removed <- read.table("T:\\DCEG\\CGF\\TechTransfer\\Microbiome\\Extraction\\Optimization\\Fresh Fecal Optimization_2017.08\\Analysis\\Run1&2\\samples_being_ignored.txt",sep="\t")

#Remove failed samples from info database
filtered_info <-subset(complete_info, !(NephID %in% complete_removed$V1))

#Verify the numbers are the same
nrow(complete_info) - nrow(complete_removed)
nrow(filtered_info)

#Remove Study samples from database
control_info <- subset(filtered_info,!Type=='Study')

#Verify the subsetting
nrow(filtered_info) - sum(filtered_info$Type=='Study')
nrow(control_info)
```

Nephele OTU Table
```{r}
#Read in BIOM database output from Nephele and convert it to a data frame
dat <- read_biom("T:\\DCEG\\CGF\\TechTransfer\\Microbiome\\Extraction\\Optimization\\Fresh Fecal Optimization_2017.08\\Analysis\\Run1&2\\otu_table_mc2_w_tax.biom")
otu_table  <- as.data.frame(as.matrix(biom_data(dat)))
nrow(otu_table)

#Subset the OTU table to include only passed control samples by finding selecting only the rows with names that 
#match the SeqID names of the control_info database
otu_table <- otu_table[,colnames(otu_table) %in% control_info$NephID]
head(otu_table)
nrow(otu_table)
```

Generate OTU reference database at the species level
```{r}
#Read in reference file
references_complete <- observation_metadata(dat)
head(references_complete)
nrow(references_complete)

#Remove dat file
remove(dat)

#Create a list of unique reference species
species_list_reference<- unique(references_complete$taxonomy7)
species_list_reference

#Crete lists of species that match both databases, calculate species missing
included_species <- species_list_controls[(species_list_controls %in% species_list_reference)]
length(species_list_controls) - length(included_species) #Number not included

#Create taxonomy database of only the included species
references_included_species <- subset(references_complete, taxonomy7 %in% included_species)
references_included_species
```
Create database of all otus for control type
```{r}
#Filter otu database (columns) by art.col.samples list (rows) to include only samples in question
art.col.complete.otu <- otu_table[,colnames(otu_table) %in% art.col.samples]
head(art.col.complete.otu)
nrow(art.col.complete.otu)

head(references_complete)
nrow(references_complete)

#Merge control otu database (rows taxonomy IDS) with control reference database (rows taxonomy IDS)
#Add rownames back to database, and remove column of rownames
art.col.complete.final.otu <- merge(art.col.complete.otu,references_complete, by="row.names")
rownames(art.col.complete.final.otu) <- art.col.complete.final.otu$Row.names
art.col.complete.final.otu <- art.col.complete.final.otu[, -which(names(art.col.complete.final.otu) %in% c("Row.names"))]
head(art.col.complete.final.otu)
nrow(art.col.complete.final.otu)

```

Create merged reference, otu, info, sample databases for species
```{r}
#Filter reference species list (found in column named Taxonomy 7) on control composition (Species column of art.col databse) to create taxonomy database containing only the species related to the control
art.col.ref <- subset(references_included_species, taxonomy7 %in% art.col$Species)
head(art.col.ref)
nrow(art.col.ref)

#Filter otu database by taxonomy row names (taxonomy IDS) found in each control (taxonomy IDS)
art.col.otu <- subset(otu_table, rownames(otu_table) %in% rownames(art.col.ref))
head(art.col.otu)
nrow(art.col.otu)

#Create list of controls by sample name
art.col.samples <- subset(control_info, Type=='AC')
art.col.samples <- art.col.samples$NephID

#Filter otu database (columns) by art.col.samples list (rows) to include only samples in question
art.col.otu <- art.col.otu[,colnames(art.col.otu) %in% art.col.samples]
head(art.col.otu)
nrow(art.col.otu)

#Filter otu database to include only samples within control type
art.col.species <- subset(art.col.otu, rownames(art.col.otu) %in% rownames(art.col.ref))
head(art.col.species)
nrow(art.col.species)

#Merge control otu database (rows taxonomy IDS) with control reference database (rows taxonomy IDS)
#Add rownames back to database, and remove column of rownames
art.col.final <- merge(art.col.species,art.col.ref, by="row.names")
rownames(art.col.final) <- art.col.final$Row.names
art.col.final <- art.col.final[, -which(names(art.col.final) %in% c("Row.names"))]
head(art.col.final)
nrow(art.col.final)
```


Save Species files 
```{r}
write.csv(art.col.final, file="artifical.colony.species.csv")
write.csv(art.col.complete.final.otu, file="artificial.colony.allotu.csv")


```

Generate OTU reference database at the genus level
```{r}
#Create a list of unique reference genus
genus_list_reference<- unique(references_complete$taxonomy7)
genus_list_reference

#Crete lists of genus that match both databases, and a list of genus not included
missing_genus <- genus_list_controls[!(genus_list_controls %in% genus_list_reference)]
missing_genus

included_genus <- genus_list_controls[(genus_list_controls %in% genus_list_reference)]
included_genus

#Create taxonomy database of only the included genus
references_included_genus <- subset(references_complete, taxonomy7 %in% included_genus)
references_included_genus
```

Not Used
```{r}
#Subset info database with only control information
control_removed 
control_otu 

#Remove the D's
#references_complete <- gsub("D_6__","",references_complete)
#references_complete <- gsub("D_5__","",references_complete)

#Transpose the otu_table so same names are in the rows, and species are in the columns
otu_table_flipped <- transpose(otu_table)
colnames(otu_table_flipped) <- rownames(otu_table)
rownames(otu_table_flipped) <- colnames(otu_table)


#Review database
head(otu_table)

#Samples in the BIOM database must have representative species with higher than 2 OTU sequences, otherwise they are dropped. In this instance 1 sample was lost - Sample SC326763.PC07578.A12
nrow(otu_controls)


```