---
title: "R Notebook"
output: word_document
editor_options: 
  chunk_output_type: console
---
Database information
#Databases Created
-dat: table of all the taxonomy assigned within Silva used in this study

-genus_percent: table of the taxonomy information of all controls including expected frequencies
 -> observed percentages added for both controls and kits
--genus_percent_ext: subset including only ext controls and kits, and genus related to the kits
 
-samples_failed: table of samples that did not meet QC thresholds

-variable_info: table of the variable information of all samples and controls
--filtered_info: table of all samples/controls that passed QC, filtering off of samples_failed
---control_info: subset of filtered_info, to only include all controls 
   ->genus count information added from genus_otu

-genus_otu: complete table of the OTUs for passed samples/controls by Genus
--genus_otucount_controls: count of the expected number of genus, the observed number of genus, and the total number of genus found in each control
---genus_otucount_ext_kit: subsetting genus_otucount_controls to include only the extraction controls, and expanding information by kit
--missing_annot: list of all the species not annotated in the Silva database, expected in the controls

-genus_otuabund_sample: table of the oberved percent frequency of each Genus for all samples
--genus_otuabund_control: subset of genus_otuabund_sample, grouped by controls
---genus_otuabund_ext_kit: subset of genus_otuabund_ext_kit, to include only ext controls, grouped by kit


#Load required libraries
```{r}
library("biom")
library(data.table)
```

#Taxonomy Reference - Control Annotations
Read in controls taxonomy file, grouped by Genus
```{r}
genus_percent <- read.table("T:\\DCEG\\CGF\\TechTransfer\\Microbiome\\Extraction\\Optimization\\Fresh Fecal Optimization_2017.08\\Controls\\Taxonomy\\Taxonomy_Genus.txt",sep="\t",header=TRUE)

#Add row names to genus full
genus_list <- unique(genus_percent[,"Genus"])
rownames(genus_percent) <- genus_list

#Create Control Composition Database, removing rows of species not in samples, followed by samples not of specified control type
##Artifical Colony
art.col.con <- subset(genus_percent, !(Art.Col=='NA'))
art.col.con <- art.col.con[,1:10]

#Zymo Sequencing Control
zymo.ext.con <-subset(genus_percent, !(Zymo.Ext=='NA'))
zymo.ext.con <- zymo.ext.con[,c(1:9,16)]

```

#Variable info
Read in the entire control variable database, and controls that failed threshold QC. 
Filter database removing all failed sampels, and any that are not a control
```{r}
#Read in database with variable information and list of samples that did not meet the sequencing threshold
variable_info <- read.table("T:\\DCEG\\CGF\\TechTransfer\\Microbiome\\Extraction\\Optimization\\Fresh Fecal Optimization_2017.08\\Analysis\\Run1&2\\Manifests\\sampledata.txt",sep="\t",header=TRUE)
samples_failed <- read.table("T:\\DCEG\\CGF\\TechTransfer\\Microbiome\\Extraction\\Optimization\\Fresh Fecal Optimization_2017.08\\Analysis\\Run1&2\\Manifests\\samples_being_ignored.txt",sep="\t")

#Remove failed samples from info database
filtered_info <-subset(variable_info, !(NephID %in% samples_failed$V1))

#Verify filtering - numbers should match
nrow(variable_info) - nrow(samples_failed)
nrow(filtered_info)

#Remove Study samples from database
control_info <- subset(filtered_info,!Type=='Study')
control_info <- subset(control_info,!Type=='EB')
control_info <- subset(control_info,!Type=='RG')

#Verify the subsetting - numbers should match
nrow(filtered_info) - sum(filtered_info$Type=='Study') - sum(filtered_info$Type=='EB') - sum(filtered_info$Type=='RG')
nrow(control_info)

#Add Row names to dataframe
row.names(control_info) <- control_info$NephID
```

#OTU Counts - OTU's by Genus (raw data)
Read in OTU information, and determine total Genus counts
```{r}
#Read in OTU database at Genus level
genus_otu <- read.csv("T:\\DCEG\\CGF\\TechTransfer\\Microbiome\\Extraction\\Optimization\\Fresh Fecal Optimization_2017.08\\Analysis\\Run1&2\\TaxaCounts\\taxa_counts_Genus.csv", row.names = 1)

#Create a total row for each sample
sample_list <- colnames(genus_otu)
genus_list <- rownames(genus_otu)

#For each sample
for (a in sample_list){
 #Set sum to 0
 sum=0
 
 #For each genus
 for (b in genus_list){
  #Add the number of counts together
  sum <- genus_otu[b,a] + sum
 }
 
 #When complete, push to a "Total_Reads" row
 genus_otu["Total_Reads",a] <- sum
 
 
}
```

#OTU Counts - Analysis Genus Expected in Controls by Control Type
Read in OTU information, and determine Genus counts (presence/absence) for each sample
```{r}
#Create control sample list, control type list
sample_list <- unique(control_info$NephID)
control_types <- unique(control_info$Type)

#Create a missing list for non-annotated Genus
missing_anno <- list()

#Determine which genus_list to use
##Run through each sample
for (a in sample_list){
 
 #Set count to 0
 count_present = 0
 count_absent = 0
 
 #Create genus list of control types
 if(control_info[a,"Type"]=="AC"){
  genus_list <- unique(art.col.con$Genus)
  } else if(control_info[a,"Type"]=="Zymo.Ext"){
   genus_list <- unique(zymo.ext.con$Genus)
  } else {genus_list <- data.frame()}
 
 #Convert to character
 genus_list <- as.character(genus_list)
 
 #For each genus is list, determine number of reads
 for (b in genus_list){
  value <- genus_otu[b,a]

  #If value is NA (meaning that the genus was not found in sample), skip
  #If reads are >0 then count_presence increases by 1
  #If reads are =0 then count_absence increases by 1 

  if(is.na(value) | is.null(value)) {
   missing_anno[[b]] <- "No Annotations"
   next
  } else if (value>0) {
   count_present = count_present +1
  } else if (value==0){
   count_absent = count_absent +1
  } else{
   next
   }
 }
 
 #Add final count and length of the list to Control_Info dataframe
 control_info[a,"GenusCount"] <- count_present
 control_info[a,"GenusAbsent"] <- count_absent
 control_info[a,"GenusExpect"] <- length(genus_list)
}

#New Database to store final Genus information
genus_otucount_controls <- data.frame()

#For each of the control types
for (b in control_types){
 
 #Reset counts
 genus=0
 total=0
 count=0
 
 #For each of the samples in the sample list
 for(a in sample_list){
  
  #Create control type groups
  valid <- control_info[a,"Type"]
  
  #If the current samples control type matches the control type of the iteration
  if(valid==b){
   
   #Add the total number of genus found
   #Add the total number of genus expected, and subtract the number missing (this is necessary as some
   #genus are not found in the list, and not necessarily absent from the sample)
   genus <- genus + control_info[a,"GenusCount"]
   total <- total + control_info[a,"GenusCount"] + control_info[a,"GenusAbsent"]
   expect <- control_info[a,"GenusExpect"]
   count <- count + 1
  } else{
    next
   }
 }
 
 #Find the average of the genus and totals
 genus_average <- genus/count
 total_average <- total/count
 
 #Push the totals to the new database 
 genus_otucount_controls[1,b] <- genus_average
 genus_otucount_controls[2,b] <- total_average
 genus_otucount_controls[3,b] <- expect

 #Add Row names
 row.names(genus_otucount_controls) <- c("Genus Count","Total", "Expected")
}
```

#OTU Counts - Analysis Genus OTU Presence/Absence by Control Type by Kit (Ext Control Only)
Extraction controls by control type by kit
```{r}
#Unique Kits, removing "None" as a kit type
kit_types <- unique(control_info$Kit)
kit_types <- kit_types[!kit_types %in% "None"]

#Control Types for Extraction only - no information for Sequencing controls
control_types <- c("AC", "Zymo.Ext")

#New Database for genus level count information from kits
genus_otucount_ext_kit <- data.frame()

#Reset counters
sums =0
row_count =0
row_count2 =0

#For each of the extraction kits
for (a in kit_types){
 
 #Group by control types
 for (b in control_types) {
  
  #Reset counters
  genus =0
  total=0
  expect=0

  #Create a list of samples that are of the specified kit and control type
  sample_list<- subset(control_info, Kit==a & Type==b)
  sample_list <- unique(sample_list$NephID)

  #For each sample in the list
  for (c in sample_list){
   
   #Count the number of unique genus found
   #Count the total number of gensus matched in reference database
   #Count the number of expected genus for this sample
   genus <- genus + control_info[c,"GenusCount"]
   total <- total + control_info[c,"GenusCount"] + control_info[c,"GenusAbsent"]
   expect <- control_info[c,"GenusExpect"]
  }
  
  #Take the average of all extracted iterations of this control  with this kit of
  genus_average <- genus/length(sample_list)
  total_average <- total/length(sample_list)
  
  #Move the row counters by 1 or 2
  row_count <- sums +1
  row_count2 <- sums+2
  
  #Push information to the database
  genus_otucount_ext_kit[row_count, b] <- genus_average
  genus_otucount_ext_kit[row_count2, b] <- genus_average
 }
 
 #Increase the row counter to skip the two filled rows
 sums = sums+2
}

#Add row names to the database
row.names (genus_otucount_ext_kit) <- c("Qiagen DSPs_Count", "Qiagen DSP_Expected", "Zymo Mag Bead_Count", "Zymo MagBead_Expected", "PowerMag Soil_Count", "PowerMag Soil_Expected", "PowerMag MB_Count", "PowerMag MB_Expected", "Qiagen QIAmp_Count", "Qiagen QIAmp_Expected")  
```

#OTU Percent - Analysis Genus OTU Abundance (All Samples)
```{r}
#Create sample list
sample_list <- unique(control_info$NephID)
genus_list <- rownames(genus_otu[1:nrow(genus_otu)-1,])

#Create new table
genus_otuabund_sample = data.frame()

#Determine which genus_list to use
##Run through each sample
for (a in sample_list){
 
 for (b in genus_list){
  
  #Determine the reads for the genus, and the total reads for the sample
  reads <- genus_otu[b,a]
  total <- genus_otu["Total_Reads",a]

  #If value is NA (meaning that the genus was not found in sample), skip
  #If reads are >0 then add the read count to reads_total

  if(is.na(reads) | is.null(reads)) {
   next
  } else {
   percent <- (reads/total)
   
   #Add out count percent to the genus_otuabund_sample database
   genus_otuabund_sample[b,a] <- percent
  }
 }
}

#Output the table of all samples
write.csv(genus_otuabund_sample,"genus_otuabund_sample.csv")
```

#OTU Percent - Analysis Genus OTU Abundance by Control Type
```{r}
#Unique Types of controls and all genus
control_types <- unique(control_info$Type)
genus_list <- rownames(genus_otu[1:nrow(genus_otu)-1,])

#New Database to store final Genus information
genus_otuabund_control <- data.frame()

#For each of the control types
for (a in control_types){
 
  #For each Genus in list
  for (b in genus_list){
   
   #Reset Counter
   reads <- 0
   total <- 0
   
   #Create a list of samples that are of the specified control type
   sample_list<- subset(control_info, Type==a)
   sample_list <- unique(sample_list$NephID)
   
   #For each of the samples in the sample list
   for(c in sample_list){
    
    #Add up the total percentages from otupercent_sample table [genus,sampleID]
    reads <- genus_otu [b,c] + reads
    total <- genus_otu["Total_Reads",c] + total
   }
   
   #average of genus level by taking total number of reads of that genus divded by 
   #total number of reads for samples included
   #Push value to new dataframe
   percent <- reads / total
   genus_otuabund_control[b,a] <- percent
  }
 }

#Output to text file
write.csv(genus_otuabund_control,"genus_otuabund_control.csv")
```

#OTU Percent - Analysis Genus OTU Abundance by Control Type by Kit (Ext Control Only)
```{r}
#Unique Types of controls and all genus
kit_types <- unique(control_info$Kit)
kit_types <- kit_types[-6] #Removing the None group
genus_list <- rownames(genus_otu[1:nrow(genus_otu)-1,])

#Store column number
cols <- ncol(genus_percent)

#For each of the kits types
for (a in kit_types){
 
  #For each Genus in list
  for (b in genus_list){
   
   #Reset Counter
   reads <- 0
   total <- 0
   
   #Create a list of samples that are of the specified control type
   sample_list<- subset(control_info, Kit==a)
   sample_list <- unique(sample_list$NephID)
   
   #For each of the samples in the sample list
   for(c in sample_list){
    
    #Add up the total percentages from otupercent_sample table [genus,sampleID]
    reads <- genus_otu [b,c] + reads
    total <- genus_otu["Total_Reads",c] + total
   }
   
   #average of genus level by taking total number of reads of that genus divded by 
   #total number of reads for samples included
   #Push value to new dataframe
   percent <- reads / total
   print(a)
   genus_percent[b,a] <- percent
  }
}
#Repeat at the control by kit level
control_types<-c("AC","Zymo.Ext")

i = ncol(genus_percent) +1
#For each of the kits types
for (a in kit_types){
 for(b in control_types){
  #Create a list of samples that are of the specified control type
  sample_list<- subset(control_info, Kit==a & Type==b)
  sample_list <- unique(sample_list$NephID)
  
  for(c in genus_list){
   
   #Reset Counter
   reads <- 0
   total <- 0

    #For each of the samples in the sample list
   for(d in sample_list){
    
    #Add up the total percentages from otupercent_sample table [genus,sampleID]
    reads <- genus_otu [c,d] + reads
    total <- genus_otu["Total_Reads",d] + total
   }
   
   #average of genus level by taking total number of reads of that genus divded by 
   #total number of reads for samples included
   #Push value to new dataframe
   percent <- reads / total
   genus_percent[c,i] <- percent
  }
  i=i+1
 }
}

i=cols+1
#Add column names
new_colnames<- c("Qiagen DSP Virus_OB", "Zymo MagBead DNone Kit_OB","MagAttract PowerMag Soil_OB",
                 "MagAttract PowerMag Microbiome_OB","Qiagen QIAamp  Modified_OB",
                 "Qiagen DSP Virus_OB_AC", "Qiagen DSP Virus_OB_ZE",
                 "Zymo MagBead DNone Kit_OB_AC","Zymo MagBead DNone Kit_OB_ZE",
                 "MagAttract PowerMag Soil_OB_AC","MagAttract PowerMag Soil_OB_ZE",
                 "MagAttract PowerMag Microbiome_OB_AC","MagAttract PowerMag Microbiome_OB_ZE",
                 "Qiagen QIAamp  Modified_OB_AC","Qiagen QIAamp  Modified_OB_ZE")

#for each header name, apply it to the recently created columns
for (a in new_colnames){
 colnames(genus_percent)[i]<- a
 i=i+1
}

#Output to text file
write.csv(genus_percent,"genus_outabund_kit.csv")
```

#OTU Percent - Analysis of Genus Percent Abundance Differences, expected versus observed by kit (Ext Only)
Perform AC and Zymo.Ext Comparisons
```{r}
#Create new genus databse with only AC and Zymo Columns
#Remove genus groups not included in control types
temp <- data.frame()
temp <- genus_percent[,c(11,16,23:32)]
temp <- subset(temp, !(Art.Col=="NA" & Zymo.Ext=="NA"))

#Save the column start
cols=ncol(genus_percent)

#Set the counter
i=cols+1

#Create lists to compare kits
genus_list <- unique(rownames(temp))
compare_list <- colnames(temp[,3:12])

#For each kit 
for (a in compare_list){
 #For each genus, within the artifical colony list
 for (b in genus_list){
  #Determine abundance value
  if(i%%2==0){
   value <- temp[b,"Zymo.Ext"]
  }
  else{value <- temp[b,"Art.Col"]
  }
  
  #If the value is NA fill in NA
  if(is.na(value)){
   genus_percent[b,i] <- "NA"
  }
  #Otherwise, fill in the abundance rate
  else{
   genus_percent[b,i] <- ((value - temp[b,a])/value)*100*-1
  }
  
 }
 print(i)
 #increase the counter
 i=i+1
}

#Add column names
new_colnames <- c("Qiagen DSP-AC_Diff","Qiagen DSP-ZE_Diff",
                  "Zymo MagBead-AC_Diff","Zymo MagBead-ZE_Diff", "MagAttract PS-AC_Diff",
                  "MagAttract PS-ZE_Diff", "MagAttract MB-AC_Diff","MagAttract MB-ZE_Diff", 
                  "Qiagen QIAmp-AC_Diff", "Qiagen QIAMmp-ZE_Diff")

i=cols+1
for (a in new_colnames){
 colnames(genus_percent)[i]<- a
 i=i+1
}

#subset table to include only relevant genus, and only control information for extraction
temp <- genus_percent[,c(11,16,19:42)]
temp <- subset(temp, !(Art.Col=="NA" & Zymo.Ext=="NA"))

#Output file
write.csv(temp,"genus_otuabund_obsVSexp_ext.csv")
```

#OTU Percent - Not expected Genus by Control Type by Kit (Ext Control Only)
```{r}

#Create temp dataframe from genus_percent
temp <- genus_percent[,c(11,16,23:32)]

#Subset by those Genus not expected to be in extraction
temp <- subset(temp, is.na(Art.Col) | is.na(Zymo.Ext))

#Output the unexpected percent frequency database
write.csv(temp,"genus_other.csv")
```
