---
title: "R Notebook"
output: word_document
---
#Load required libraries
```{r}
library("biom")
library(data.table)
```
#Reference Silva Database
Nephele OTU Table
```{r}
#Read in BIOM database output from Nephele and convert it to a data frame
dat <- read.table("T:\\DCEG\\CGF\\Laboratory\\Projects\\MR-0084\\NP0084-MB5\\QC Data\\qiime\\PipelineResults_N1O8V48GTJ74\\otus\\sortmerna_assigned_taxonomy\\rep_set_tax_assignments.txt",
                  sep="\t")
otu_table  <- as.data.frame(as.matrix(biom_data(dat)))
```


#Control Taxonomy - Genus
Read in controls taxonomy file, grouped by Genus
```{r}
genus_full <- read.table("T:\\DCEG\\CGF\\TechTransfer\\Microbiome\\Extraction\\Optimization\\Fresh Fecal Optimization_2017.08\\Controls\\Taxonomy\\Taxonomy_Genus.txt",sep="\t",header=TRUE)

#Create Control Composition Database, removing rows of species not in samples, followed by samples not of specified control type
##Artifical Colony
art.col.con <- subset(genus_full, !(Art.Col=='NA'))
art.col.con <- art.col.con[,1:10]

##Two concentrations of zymo extraction controls
zymo.seq200.con <- subset(genus_full, !(Zymo.Ext=='NA'))
zymo.seq200.con <- zymo.seq200.con[,c(1:9,15)]
zymo.seq2000.con <- zymo.seq200.con

#Zymo Sequencing Control
zymo.ext.con <-subset(genus_full, !(Zymo.Seq=='NA'))
zymo.ext.con <- zymo.ext.con[,c(1:9,16)]

#ATTC Controls
msa1000.con <-subset(genus_full, !(MSA1000=='NA'))
msa1000.con <- msa1000.con[,c(1:9,11)]

msa1001.con <- subset(genus_full, !(MSA1001=='NA'))
msa1001.con <- msa1001.con[,c(1:9,12)]

msa1002.con <- subset(genus_full, !(MSA1002=='NA'))
msa1002.con <- msa1002.con[,c(1:9,13)]

msa1003.con <- subset(genus_full, !(MSA1003=='NA'))
msa1003.con <- msa1003.con[,c(1:9,14)]

```

#Variable info
Read in the entire sample/control info database, and samples/controls that failed threshold QC. Filter database by not passing, and by not control
```{r}
#Read in database with variable information and list of samples that did not meet the sequencing threshold
variable_info <- read.table("T:\\DCEG\\CGF\\TechTransfer\\Microbiome\\Extraction\\Optimization\\Fresh Fecal Optimization_2017.08\\Analysis\\Run1&2\\Manifests\\sampledata.txt",sep="\t",header=TRUE)
samples_failed <- read.table("T:\\DCEG\\CGF\\TechTransfer\\Microbiome\\Extraction\\Optimization\\Fresh Fecal Optimization_2017.08\\Analysis\\Run1&2\\Manifests\\samples_being_ignored.txt",sep="\t")

#Remove failed samples from info database
filtered_info <-subset(variable_info, !(NephID %in% samples_failed$V1))

#Verify the numbers are the same
nrow(variable_info) - nrow(samples_failed)
nrow(filtered_info)

#Remove Study samples from database
control_info <- subset(filtered_info,!Type=='Study')
control_info <- subset(control_info,!Type=='EB')
control_info <- subset(control_info,!Type=='RG')

#Verify the subsetting
nrow(filtered_info) - sum(filtered_info$Type=='Study') - sum(filtered_info$Type=='EB') - sum(filtered_info$Type=='RG')
nrow(control_info)

#Add Row names to dataframe
row.names(control_info) <- control_info$NephID

```
#OTU Info
Read in OTU database at Genus level
```{r}
genus_otu <- read.csv("T:\\DCEG\\CGF\\TechTransfer\\Microbiome\\Extraction\\Optimization\\Fresh Fecal Optimization_2017.08\\Analysis\\Run1&2\\TaxaCounts\\taxa_counts_Genus_edited.csv", row.names = 1)
```

Genus OTU count by sample - Add to Control_info table
```{r}
#Create sample list
sample_list <- unique(control_info$NephID)
sample_list <- as.character(sample_list)

#Change the vector type
control_info$Type <- as.character(control_info$Type)
control_info$Kit <- as.character(control_info$Kit)

#Determine which genus_list to use
##Run through each sample
for (a in sample_list){
 
 #Set count to 0
 count_present = 0
 count_absent = 0
 
 #Create genus list of control types
 if(control_info[a,"Type"]=="AC"){
  genus_list <- unique(art.col.con$Genus)
  } else if(control_info[a,"Type"]=="Zymo.Ext"){
   genus_list <- unique(zymo.ext.con$Genus)
  } else if(control_info[a,"Type"]=="MSA-1000"){
    genus_list <- unique(msa1000.con$Genus)
  } else if(control_info[a,"Type"]=="MSA-1001"){
   genus_list <- unique(msa1001.con$Genus)
   } else if(control_info[a,"Type"]=="MSA-1002"){
    genus_list <- unique(msa1002.con$Genus)
    } else if(control_info[a,"Type"]=="MSA-1003"){
     genus_list <- unique(msa1003.con$Genus)
     } else if(control_info[a,"Type"]=="Zymo.Seq.200"){
      genus_list <- unique(zymo.seq200.con$Genus)
      } else if(control_info[a,"Type"]=="Zymo.Seq.2000"){
       genus_list <- unique(zymo.seq2000.con$Genus)
       } else {genus_list <- data.frame()}
 
 #Convert to character
 genus_list <- as.character(genus_list)
 
 #For each genus is list, determine number of reads
 for (b in genus_list){
  value <- genus_otu[b,a]

  #If value is NA (meaning that the genus was not found in sample), skip
  #If reads are >0 then count_presence increases by 1
  #If reads are =0 then count_absence increases by 1 

  if(is.na(value) | is.null(value)) {
   next
  } else if (value>0) {
   count_present = count_present +1
  } else if (value==0){
   count_absent = count_absent +1
  } else{
   next
   }
 }
 
 #Add final count and length of the list to Control_Info dataframe
 control_info[a,"GenusCount"] <- count_present
 control_info[a,"GenusAbsent"] <- count_absent
 control_info[a,"GenusExpect"] <- length(genus_list)
}
```

Analysis Genus OTU Count - All controls by control type
```{r}
#Convert column to numeric
control_info$GenusCount <- as.numeric(control_info$GenusCount)
control_info$GenusExpect <- as.numeric(control_info$GenusExpect)
control_info$GenusAbsent <- as.numeric(control_info$GenusAbsent)

#Unique Types of controls
control_types <- unique(control_info$Type)

#New Database to store final Genus information
genus_otucount <- data.frame()

#For each of the control types
for (b in control_types){
 
 #Reset counts
 genus=0
 total=0
 count=0
 
 #For each of the samples in the sample list
 for(a in sample_list){
  
  #Create control type groups
  valid <- control_info[a,"Type"]
  
  #If the current samples control type matches the control type of the iteration
  if(valid==b){
   
   #Add the total number of genus found
   #Add the total number of genus expected, and subtract the number missing (this is necessary as some
   #genus are not found in the list, and not necessarily absent from the sample)
   genus <- genus + control_info[a,"GenusCount"]
   total <- total + control_info[a,"GenusCount"] + control_info[a,"GenusAbsent"]
   expect <- control_info[a,"GenusExpect"]
   count <- count + 1
  } else{
    next
   }
 }
 
 #Find the average of the genus and totals
 genus_average <- genus/count
 total_average <- total/count
 
 #Push the totals to the new database 
 genus_otucount[1,b] <- genus_average
 genus_otucount[2,b] <- total_average
 genus_otucount[3,b] <- expect

 #Add Row names
 row.names(genus_otucount) <- c("Genus Count","Total", "Expected")
}
```

Analysis Genus OTU Count - Extraction controls by control type by kit
```{r}
#Unique Kits, removing "None" as a kit type
kit_types <- unique(control_info$Kit)
kit_types <- kit_types[!kit_types %in% "None"]

#Control Types for Extraction only
control_types <- c("AC", "Zymo.Ext")

#New Database for genus level count information from kits
genus_otucount_kit <- data.frame()

#Reset counters
sums =0
row_count =0
row_count2 =0

#For each of the extraction kits
for (a in kit_types){
 
 #Group by control types
 for (b in control_types) {
  
  #Reset counters
  genus =0
  total=0
  expect=0

  #Create a list of samples that are of the specified kit and control type
  sample_list<- subset(control_info, Kit==a & Type==b)
  sample_list <- unique(sample_list$NephID)
  sample_list <- as.character(sample_list)
  
  #For each sample in the list
  for (c in sample_list){
   
   #Count the number of unique genus found
   #Count the total number of gensus matched in reference database
   #Count the number of expected genus for this sample
   genus <- genus + control_info[c,"GenusCount"]
   total <- total + control_info[c,"GenusCount"] + control_info[c,"GenusAbsent"]
   expect <- control_info[c,"GenusExpect"]
  }
  
  #Take the average of all extracted iterations of this control  with this kit of
  genus_average <- genus/length(sample_list)
  total_average <- total/length(sample_list)
  
  #Move the row counters by 1 or 2
  row_count <- sums +1
  row_count2 <- sums+2
  
  #Push information to the database
  genus_otucount_kit[row_count, b] <- genus_average
  genus_otucount_kit[row_count2, b] <- genus_average
 }
 
 #Increase the row counter to skip the two filled rows
 sums = sums+2
}

#Add row names to the database
row.names (genus_otucount_kit) <- c("Qiagen DSPs_Count", "Qiagen DSP_Expected", "Zymo Mag Bead_Count", "Zymo MagBead_Expected", "PowerMag Soil_Count", "PowerMag Soil_Expected", "PowerMag MB_Count", "PowerMag MB_Expected", "Qiagen QIAmp_Count", "Qiagen QIAmp_Expected")  
```

Genus OTU % by sample
```{r}
#Create sample list
sample_list <- unique(control_info$NephID)
sample_list <- as.character(sample_list)

#Change the vector type
control_info$Type <- as.character(control_info$Type)
control_info$Kit <- as.character(control_info$Kit)

#Create new table
genus_otupercent_sample = data.frame()

#Determine which genus_list to use
##Run through each sample
for (a in sample_list){
 
 #Set count to 0
 percent = 0

 #Create genus list of control types
 if(control_info[a,"Type"]=="AC"){
  genus_list <- unique(art.col.con$Genus)
  } else if(control_info[a,"Type"]=="Zymo.Ext"){
   genus_list <- unique(zymo.ext.con$Genus)
  } else if(control_info[a,"Type"]=="MSA-1000"){
    genus_list <- unique(msa1000.con$Genus)
  } else if(control_info[a,"Type"]=="MSA-1001"){
   genus_list <- unique(msa1001.con$Genus)
   } else if(control_info[a,"Type"]=="MSA-1002"){
    genus_list <- unique(msa1002.con$Genus)
    } else if(control_info[a,"Type"]=="MSA-1003"){
     genus_list <- unique(msa1003.con$Genus)
     } else if(control_info[a,"Type"]=="Zymo.Seq.200"){
      genus_list <- unique(zymo.seq200.con$Genus)
      } else if(control_info[a,"Type"]=="Zymo.Seq.2000"){
       genus_list <- unique(zymo.seq2000.con$Genus)
       } else {genus_list <- data.frame()}
 
 #Convert genus_list to character
 genus_list <- as.character(genus_list)

 #Sum all the read numbers
 total <- sum(genus_otu[,a])

 #For each genus is list, determine number of reads
 for (b in genus_list){
  value <- genus_otu[b,a]

  #If value is NA (meaning that the genus was not found in sample), skip
  #If reads are >0 then add the read count to reads_total

  if(is.na(value) | is.null(value)) {
   next
  } else if (value>0) {
   percent <- value/total
   
   #Add out count percent to the genus_otupercent_sample database
   genus_otupercent_sample[b,a] <- percent
  } else{
   next
  }
 }
}

write.csv(genus_otupercent_sample,"genus_percent.csv")

```

Analysis Genus Percent - All control by control type
```{r}
#Unique Types of controls
control_info$Type <- as.character(control_info$Type)
control_types <- unique(control_info$Type)

#New Database to store final Genus information
genus_otupercent_control <- data.frame()

#For each of the control types
for (a in control_types){
 
  #Create a list of samples that are of the specified control type
  sample_list<- subset(control_info, Type==a)
  sample_list <- unique(sample_list$NephID)
  sample_list <- as.character(sample_list)
  
  #Create genus list of control types
  if(a=="AC"){
   genus_list <- unique(art.col.con$Genus)
   } else if(a=="Zymo.Ext"){
    genus_list <- unique(zymo.ext.con$Genus)
   } else if(a=="MSA-1000"){
     genus_list <- unique(msa1000.con$Genus)
   } else if(a=="MSA-1001"){
    genus_list <- unique(msa1001.con$Genus)
    } else if(a=="MSA-1002"){
     genus_list <- unique(msa1002.con$Genus)
     } else if(a=="MSA-1003"){
      genus_list <- unique(msa1003.con$Genus)
      } else if(a=="Zymo.Seq.200"){
       genus_list <- unique(zymo.seq200.con$Genus)
       } else if(a=="Zymo.Seq.2000"){
        genus_list <- unique(zymo.seq2000.con$Genus)
       } else {next}
  
  #For each Genus in list
  for (b in genus_list){
   
   #Reset Counter
   total <- 0
   
   #For each of the samples in the sample list
   for(c in sample_list){
    
    #Add up the total percentages from otupercent_sample table [genus,sampleID]
    total <- genus_otupercent_sample [b,c] + total
   
   }
   
   #average of genus level by taking total / number of samples
   #Push value to new dataframe
   percent <- total / length(sample_list)
   genus_otupercent_control[b,a] <- percent
  }
 }

#Output to text file
write.csv(genus_otupercent_control,"output.csv")

```

Analysis Genus Percent - Extraction control by control type by kit
```{r}
#Unique Kits, removing "None" as a kit type
kit_types <- unique(control_info$Kit)
kit_types <- kit_types[!kit_types %in% "None"]

#Control Types for Extraction only
control_types <- c("AC", "Zymo.Ext")

#New Database to store final Genus information
genus_otupercent_kit <- data.frame()

#Start Counter
count = 0

#For each kit
for (d in kit_types){
 
 #For each of the control types
 for (a in control_types){
  
  #Counter
  count = count + 1
  print (count)
  print (a)
  print (d)
  
   #Create a list of samples that are of the specified control type
   sample_list<- subset(control_info, Type==a & Kit==d)
   sample_list <- unique(sample_list$NephID)
   sample_list <- as.character(sample_list)
   
   #Create genus list of control types
   if(a=="AC"){
    genus_list <- unique(art.col.con$Genus)
    } else if(a=="Zymo.Ext"){
     genus_list <- unique(zymo.ext.con$Genus)
    } else {next}
   
   #For each Genus in list
   for (b in genus_list){
    
    #Reset Counter
    total <- 0
    
    #For each of the samples in the sample list
    for(c in sample_list){
     
     #Add up the total percentages from otupercent_sample table [genus,sampleID]
     total <- genus_otupercent_sample [b,c] + total
    
    }
    
    #average of genus level by taking total / number of samples
    #Push value to new dataframe
    percent <- total / length(sample_list)
    genus_otupercent_kit[b,count] <- percent
   }
  }
}

#Add column names
colnames(genus_otupercent_kit) <- c("Qiagen DSP-AC", "Qiagen DSP-ZE", "Zymo MagBead-AC", "Zymo MagBead-ZE", "MagAttract PS-AC", "MagAttract PS-ZE", "MagAttract MB-AC", "MagAttract MB-ZE", "Qiagen QIAmp-AC", "Qiagen QIAMmp-ZE")

#Output to text file
write.csv(genus_otupercent_kit,"genus_otupercent_kit.csv")
```

Add Percent Genus OTU to Control_info database
```{r}
genus_list <- unique(genus_full[,"Genus"])
rownames(genus_full) <- genus_list

kit_controls <- colnames(genus_otupercent_kit)
genus_list <- rownames(genus_otupercent_kit)

for (a in kit_controls){
 
 for (b in genus_list){
  
  genus_full[b,a] <-genus_otupercent_kit[b,a]
 }
}

write.csv(genus_full,"output.csv")

```

Perform AC and Zymo.Ext Comparisons
```{r}

#Create new genus databse with only AC and Zymo Columns
genus_compare <- genus_full[,c("Art.Col", "Zymo.Ext","Qiagen DSP-AC","Qiagen DSP-ZE","Zymo MagBead-AC","Zymo MagBead-ZE","MagAttract PS-AC","MagAttract PS-ZE","MagAttract MB-AC","MagAttract MB-ZE","Qiagen QIAmp-AC","Qiagen QIAMmp-ZE")]

#Remove genus groups not included in control types
genus_compare <- subset(genus_compare, !(Art.Col=="NA" & Zymo.Ext=="NA"))
genus_list <- unique(rownames(genus_compare))

#Generate column name of kit controls
kit_controls <- colnames(genus_compare[,3:12])

#Start adding new columns after current n of columns
count <- ncol(genus_compare)

#Determine percent differences
##For each of the kit controls
for (a in kit_controls){
 #Start the column is the next available position
 count = count +1

 #Move through each genus (row)
 for (b in genus_list){
  #If the column is even, it's a Zymo.Ext
  if(count %% 2 ==0){
   #If there was nothing detected for the genus, count it as 0
   if(is.na(genus_compare[b,a])){
    #If the genus is included, but should have a count of 0, set value to 0
    if(is.na(genus_compare[b,2]==0)){
     value <- ((genus_compare[b,2] - 0)/(genus_compare[b,2]))*100
     genus_compare[b,count] <- value
    }
    else {
     print (b)
     value <- 0
     genus_compare[b,count] <- value
     } 
   }
   #If genus were detected, determine the percent difference from expected
   else{
    value <- ((genus_compare[b,2] - genus_compare[b,a])/(genus_compare[b,2]))*100
    genus_compare[b,count] <- value
    }
  }
  #If the column is odd, it's an AC
  else{
   #If there was nothing detected for the genus, count it as 0
   if(is.na(genus_compare[b,a])|(genus_compare[b,a]==0) ){
    value <- ((genus_compare[b,1] - 0)/(genus_compare[b,1]))*100
    genus_compare[b,count] <- value
   }
   #If genus were detected, determine the percent difference from expected
   else{
   value <- ((genus_compare[b,1] - genus_compare[b,a])/(genus_compare[b,1]))*100
   genus_compare[b,count] <- value
   }
  }
 }
}

#Add column names
colnames(genus_compare) <- c( "Art.Col","Zymo.Ext","Qiagen DSP-AC","Qiagen DSP-ZE","Zymo MagBead-AC","Zymo MagBead-ZE","MagAttract PS-AC","MagAttract PS-ZE","MagAttract MB-AC","MagAttract MB-ZE","Qiagen QIAmp-AC","Qiagen QIAMmp-ZE","Qiagen DSP-AC_Diff","Qiagen DSP-ZE_Diff","Zymo MagBead-AC_Diff","Zymo MagBead-ZE_Diff","MagAttract PS-AC_Diff","MagAttract PS-ZE_Diff","MagAttract MB-AC_Diff","MagAttract MB-ZE_Diff","Qiagen QIAmp-AC_Diff","Qiagen QIAMmp-ZE")

write.csv(genus_compare,"output.csv")

```

Missing Species
```{r}
#Determine if missing species are due to annotation problem, or simply not extracted
genus_list <- unique(rownames(genus_compare))
missing <- list()

#For each column
for (a in colnames(genus_compare)){
 #For each of the genus
 for (b in genus_list){
  #Find the percent difference
  tempval <- as.numeric(genus_compare[b,a])

  #If there is a null value
  if(is.na(tempval)) {next}
  #If the value is equal to 100 or 0
  else if(tempval==100 | tempval==0){
   #Then search for the row which contains the complete Silva annotation name
   row <- grep(b, genus_full$Naming)
   #Then determine the fill name
   temp <- as.character(genus_full[row,"Naming"])
   #Then search for the value in the reference database
   temp <- grep(temp,dat$V2)

   if(!(length(temp)>0)){
    missing[[b]] <- "No Annotations"
   } else{next}
   }
  }
}

#Print missing annotation list
print (missing)
```

Other Genus
```{r}



```