---
title: "R Notebook"
output: word_document
editor_options: 
  chunk_output_type: console
---
#Read in libraries
```{r}
library(ggplot2)
```

#Create Control database
```{r}
genus_percent <- read.table("T:\\DCEG\\CGF\\TechTransfer\\Microbiome\\Extraction\\Optimization\\Fresh Fecal Optimization_2017.08\\Controls\\Taxonomy\\Taxonomy_Genus.txt",sep="\t",header=TRUE)
rownames(genus_percent)<-genus_percent$Genus
head(genus_percent)
```

Expected Data
#Create Nephele / QIIME2 table
```{r}
#Read in tables
neph <- read.csv("Data/genus_otuabund_seq_neph_obsVexp_tab.csv",header=TRUE,stringsAsFactors = FALSE)
q2 <- read.csv("Data/genus_otuabund_seq_q2_obsVexp_tab.csv",header=TRUE,stringsAsFactors = FALSE)

#Add analysis tool as a variable
for (i in 1:nrow(neph)){
 neph[i,"BINF"]<- "Nephele"
}

for (i in 1:nrow(q2)){
 q2[i,"BINF"]<-"QIIME2"
}

#Remove col of numbers
neph <- neph[,-1]
q2 <- q2[,-1]

#Review
head(neph)
head(q2)

#Combine the tables
##Initialize counts
complete <- neph
rows = nrow(complete)+1
#Iterate through table
for (i in 1:nrow(q2)){
 complete[rows,"Genus"]<-q2[i,1]
 complete[rows,"value"]<-q2[i,2]
 complete[rows,"Plate"]<-q2[i,3]
 complete[rows,"Type"]<-q2[i,4]
 complete[rows,"BINF"]<-q2[i,5]
 rows=rows+1
}
#Add in the expected values to each column
for (i in 1:nrow(complete)){
 type <- complete[i,"Type"]
 genus <- complete[i,"Genus"]

 #Account for Seq controls being identicallllll
 if(type=="Seq.200" | type=="Seq.2000"){
  type= "Zymo.Seq"
 }

 complete[i,"Expected"]<- genus_percent[genus,type]
}
#Preview data
head(complete)
tail(complete)
```

#Create averaged dataset - Seq200 and Seq2000
```{r}
#Subset datatable to include only sequencing controls
complete_seq <- subset(complete,Type=="Seq.200" | Type=="Seq.2000")
head(complete_seq)

#Create list of genus
genus_list <- unique(complete_seq$Genus)

#Create new dataframe for averages
average_seq <- data.frame()

for (a in genus_list){
 neph_val=0; count_neph=0
 q2_val=0; count_q=0
 
 for (b in 1:nrow(complete_seq)){
  genus <- complete_seq[b,"Genus"]
  value <- complete_seq[b,"value"]
  binf <- complete_seq[b,"BINF"]

  if(is.na(value)|is.nan(value)){
   next
  } else{
   
   if(a==genus){
    
    if(binf=="Nephele"){
    neph_val <- neph_val + value
    count_neph <- count_neph +1
    } else{
     q2_val <- q2_val + value
     count_q <- count_q + 1
    }
   }
  }
 }
 average_seq[a,"Genus"]<-a
 average_seq[a,"ValueQ"]<-q2_val/count_q
 average_seq[a,"ValueN"]<-neph_val/count_neph
}

#Remove all NA values
average_seq<-subset(average_seq,!(is.na(ValueQ))|!(is.na(ValueN)))

#Review results
head(average_seq)
nrow(average_seq)


#Plot graphs
ggplot(data=average_seq,mapping=aes(x=ValueQ,y=ValueN,jitter(Expected,factor=1)))+
 geom_line()+ #Can add points to graph
 geom_smooth(method = "lm",se=FALSE)+ #Add a smooth line
 labs(title="Zymo Seq Controls Compared - Expected", x="Actual QIIME2 Value", y= "Actual Neph Value")+
 geom_text(mapping=aes(label=Genus))

#Write file
write.csv(average_seq,"Data/averages_seq_nephandq2.csv")
```

#Create averaged dataset - Seq200 ONLY
```{r}
#Subset datatable to include only sequencing controls
complete_seq <- subset(complete,Type=="Seq.200")
head(complete_seq)

#Create list of genus
genus_list <- unique(complete_seq$Genus)

#Create new dataframe for averages
average_seq <- data.frame()

for (a in genus_list){
 neph_val=0; count_neph=0
 q2_val=0; count_q=0
 
 for (b in 1:nrow(complete_seq)){
  genus <- complete_seq[b,"Genus"]
  value <- complete_seq[b,"value"]
  binf <- complete_seq[b,"BINF"]

  if(is.na(value)|is.nan(value)){
   next
  } else{
   
   if(a==genus){
    
    if(binf=="Nephele"){
    neph_val <- neph_val + value
    count_neph <- count_neph +1
    } else{
     q2_val <- q2_val + value
     count_q <- count_q + 1
    }
   }
  }
 }
 average_seq[a,"Genus"]<-a
 average_seq[a,"ValueQ"]<-q2_val/count_q
 average_seq[a,"ValueN"]<-neph_val/count_neph
}

#Remove all NA values
average_seq<-subset(average_seq,!(is.na(ValueQ))|!(is.na(ValueN)))

#Review results
head(average_seq)
nrow(average_seq)

#Plot graphs
ggplot(data=average_seq,mapping=aes(x=ValueQ,y=ValueN,jitter(Expected,factor=1)))+
 geom_line()+ #Can add points to graph
 geom_smooth(method = "lm",se=FALSE)+ #Add a smooth line
 labs(title="Zymo 200Seq Neph & Q2 Compared - Expected", x="Actual QIIME2 Value", y= "Actual Neph Value")+
 geom_text(mapping=aes(label=Genus))

#Write file
write.csv(average_seq,"Data/averages_seq_200zymo_nephandq2.csv")
```

#Create averaged dataset - Seq2000 ONLY
```{r}
#Subset datatable to include only sequencing controls
complete_seq <- subset(complete,Type=="Seq.200" | Type=="Seq.2000")
head(complete_seq)

#Create list of genus
genus_list <- unique(complete_seq$Genus)

#Create new dataframe for averages
average_seq <- data.frame()

for (a in genus_list){
 neph_val=0; count_neph=0
 q2_val=0; count_q=0
 
 for (b in 1:nrow(complete_seq)){
  genus <- complete_seq[b,"Genus"]
  value <- complete_seq[b,"value"]
  binf <- complete_seq[b,"BINF"]

  if(is.na(value)|is.nan(value)){
   next
  } else{
   
   if(a==genus){
    
    if(binf=="Nephele"){
    neph_val <- neph_val + value
    count_neph <- count_neph +1
    } else{
     q2_val <- q2_val + value
     count_q <- count_q + 1
    }
   }
  }
 }
 average_seq[a,"Genus"]<-a
 average_seq[a,"ValueQ"]<-q2_val/count_q
 average_seq[a,"ValueN"]<-neph_val/count_neph
}

#Remove all NA values
average_seq<-subset(average_seq,!(is.na(ValueQ))|!(is.na(ValueN)))

#Review results
head(average_seq)
nrow(average_seq)

#Plot graphs
ggplot(data=average_seq,mapping=aes(x=ValueQ,y=ValueN,jitter(Expected,factor=1)))+
 geom_line()+ #Can add points to graph
 geom_smooth(method = "lm",se=FALSE)+ #Add a smooth line
 labs(title="Zymo 2000Seq Neph & Q2 Compared - Expected", x="Actual QIIME2 Value", y= "Actual Neph Value")+
 geom_text(mapping=aes(label=Genus))

#Write file
write.csv(average_seq,"Data/averages_seq_2000zymo_nephandq2.csv")

```

Unexpected Data
#Create Nephele / QIIME2 table
```{r}
#Read in tables
neph <- read.csv("Data/genus_otuabund_seq_neph_other_tab.csv",header=TRUE,stringsAsFactors = FALSE)
q2 <- read.csv("Data/genus_otuabund_seq_q2_other_tab.csv",header=TRUE,stringsAsFactors = FALSE)

#Add analysis tool as a variable
for (i in 1:nrow(neph)){
 neph[i,"BINF"]<- "Nephele"
}

for (i in 1:nrow(q2)){
 q2[i,"BINF"]<-"QIIME2"
}

#Remove col of numbers
neph <- neph[,-1]
q2 <- q2[,-1]

#Review
head(neph)
head(q2)

#Combine the tables
##Initialize counts
complete <- neph
rows = nrow(complete)+1
#Iterate through table
for (i in 1:nrow(q2)){
 complete[rows,"Genus"]<-q2[i,1]
 complete[rows,"value"]<-q2[i,2]
 complete[rows,"Plate"]<-q2[i,3]
 complete[rows,"Type"]<-q2[i,4]
 complete[rows,"BINF"]<-q2[i,5]
 rows=rows+1
}
head(complete)
tail(complete)

#Add in the expected values to each column
for (i in 1:nrow(complete)){
 type <- complete[i,"Type"]
 genus <- complete[i,"Genus"]
 
 #Account for Seq controls being identical
 if(type=="Seq200" | type=="Seq2000"){
  type= "Zymo.Seq"
 }
 
 complete[i,"Expected"]<- genus_percent[genus,type]
}

#Preview data
head(complete)
tail(complete)
```

#Create averaged dataset - Seq200 and Seq2000 VS Q2 and Nephele
```{r}
#Subset datatable to include only sequencing controls
complete_seq <- subset(complete,Type=="Seq200" | Type=="Seq2000")
complete_seq <- subset(complete_seq,is.na(Expected))
head(complete_seq)

#Create list of genus
genus_list <- unique(complete_seq$Genus)

#Create new dataframe for averages
average_seq <- data.frame()

for (a in genus_list){
 neph_val200=0; count_neph200=0
 q2_val200=0; count_q200=0
 neph_val2000=0; count_neph2000=0
 q2_val2000=0; count_q2000=0
 
 for (b in 1:nrow(complete_seq)){
  genus <- complete_seq[b,"Genus"]
  value <- complete_seq[b,"value"]
  binf <- complete_seq[b,"BINF"]
  seq <- complete_seq[b,"Type"]

  if(is.na(value)|is.nan(value)|value==0){
   next
  } else{
   
   if(a==genus){
    
    if(binf=="Nephele"){
     if(seq=="Seq200"){
      neph_val200 <- neph_val200 + value
      count_neph200 <- count_neph200 +1
     } else{
      neph_val2000 <- neph_val2000 + value
      count_neph2000 <- count_neph2000 +1
     }
    } else{
     if(seq=="Seq2000"){
      q2_val200 <- q2_val200 + value
      count_q200 <- count_q200 + 1
     } else{
      q2_val2000 <- q2_val2000 + value
      count_q2000 <- count_q2000 + 1
     }

    }
   }
  }
 }
 average_seq[a,"Genus"]<-a
 average_seq[a,"ValueQ200"]<-q2_val200/count_q200
 average_seq[a,"ValueN200"]<-neph_val200/count_neph200
 average_seq[a,"ValueQ2000"]<-q2_val2000/count_q2000
 average_seq[a,"ValueN2000"]<-neph_val2000/count_neph2000
}

colnames(average_seq)

#Create tableau file
temp<-data.frame()
binf <- c("","QIIME2","Nephele","QIIME2","Nephele")
control <- c("","Zymo.200","Zymo.200","Zymo.2000","Zymo.2000")
count=1

for (a in 2:5){
 for(b in 1:nrow(average_seq)){
  temp[count,"Genus"] <- rownames(average_seq[b,])
  temp[count,"value"] <- average_seq[b,a]
  temp[count,"Control"] <- control[a]
  temp[count,"BINF"] <- binf[a]
  count=count+1
 }
}
head(temp)

write.csv(temp,"Data/averages_seq_nephandq2_unexp_tab.csv")
```

#Create averaged dataset - Seq200 and Seq2000 Complete
```{r}
#Subset datatable to include only sequencing controls
complete_seq <- subset(complete,Type=="Seq200" | Type=="Seq2000")
complete_seq <- subset(complete_seq,is.na(Expected))
head(complete_seq)

#Create list of genus
genus_list <- unique(complete_seq$Genus)

#Create new dataframe for averages
average_seq <- data.frame()

for (a in genus_list){
 neph_val=0; count_neph=0
 q2_val=0; count_q=0
 
 for (b in 1:nrow(complete_seq)){
  genus <- complete_seq[b,"Genus"]
  value <- complete_seq[b,"value"]
  binf <- complete_seq[b,"BINF"]

  if(is.na(value)|is.nan(value)|value==0){
   next
  } else{
   
   if(a==genus){
    
    if(binf=="Nephele"){
    neph_val <- neph_val + value
    count_neph <- count_neph +1
    } else{
     q2_val <- q2_val + value
     count_q <- count_q + 1
    }
   }
  }
 }
 average_seq[a,"Genus"]<-a
 average_seq[a,"ValueQ"]<-q2_val/count_q
 average_seq[a,"ValueN"]<-neph_val/count_neph
}

#Remove all NA values for both datasets
average_seq<-average_seq[complete.cases(average_seq),]

#Review results
head(average_seq)
nrow(average_seq)

#Plot graphs
ggplot(data=average_seq,mapping=aes(x=ValueQ,y=ValueN,jitter(Expected,factor=1)))+
 geom_line()+ #Can add points to graph
 geom_smooth(method = "lm",se=FALSE)+ #Add a smooth line
 labs(title="Zymo Seq Controls Compared - Unexpected", x="Actual QIIME2 Value", y= "Actual Neph Value")+
 geom_text(mapping=aes(label=Genus))

#Remove Other and replot
average_seq2<-subset(average_seq,!Genus=="Other")

#Plot graphs
ggplot(data=average_seq2,mapping=aes(x=ValueQ,y=ValueN,jitter(Expected,factor=1)))+
 geom_line()+ #Can add points to graph
 geom_smooth(method = "lm",se=FALSE)+ #Add a smooth line
 labs(title="Zymo Seq Controls Compared - Unexpected", x="Actual QIIME2 Value", y= "Actual Neph Value")+
 geom_text(mapping=aes(label=Genus))

#Write file
write.csv(average_seq,"Data/averages_seq_nephandq2_unexp_complete.csv")
```